modules/lib/hyprlang.nix

```diff
@@ -0,0 +1,62 @@
+{ lib }: rec {
+  valueType = with lib.types;
+    nullOr
+    (oneOf [ bool int float str path (attrsOf valueType) (listOf valueType) ])
+    // {
+      description = "Hyprlang configuration value";
+    };
+
+  toHyprconf = with lib;
```

Comment:
It seems this contains quite a few things that are very specific to Hyprland. Consider, for example, a program other than Hyprland that uses "hyprlang" and a configuration entry "device"…

I would suggest, for example, to instead allow `toHyprconf` to either take an attribute set or a list of attribute sets as input. If the list method is chosen then the output is in that order. The Hyprland module would then be responsible for doing the device and important prefixes stuff and constructing the list accordingly. This function shouldn't have to know anything Hyprland specific.

Also, the function should be placed in `generators.nix` next to `toKDL` 🙂

Reply:
Hmm.. alright, I'll look into that... If we go this route of moving the generator out of the library module, does the valueType need to go back to being declared in each module.

Reply:
Apparently so. I think a Nixpkgs `pkgs/pkgs-lib/formats.nix`-style file would fare better in this context.

To be honest, I still don't understand the separation in Nixpkgs between `formatters` and `generators`. But that is off-topic.

Reply:
In implementing this... is there any actual reason for the ordering of specific sections? ie: the device configuration in the hyprland example being at the top or was it simply preference of the name attribute at the top of the attribute set?

Reply:
Update includes this change where it can function handling the list of attribute sets properly without needing to manually specify them. Also added another test so we can see it properly handling a singular device attribute set and a list of device attribute sets.

Reply:
> In implementing this... is there any actual reason for the ordering of specific sections? ie: the device configuration in the hyprland example being at the top or was it simply preference of the name attribute at the top of the attribute set? Because I can just set up an important prefix in modules for the attributes we want generated at the top of their sections, if that was the reason.

The device section can be anywhere. But the `name` field has to be the first in the section, that's how vaxry implemented the parsing.